<a href="https://colab.research.google.com/github/welliver724/danl-m1-hw1-Welliver-Justin.ipynb/blob/main/danl-m1-hw3-welliver-justin.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Below is to import the pandas library as pd
import pandas as pd

# Below is for an interactive display of DataFrame in Colab
from google.colab import data_table
data_table.enable_dataframe_formatter()

# Below is to read nba.csv as nba DataFrame
nyc_payroll = pd.read_csv('https://bcdanl.github.io/data/nyc_payroll_2022_2023.csv',
                          parse_dates = ["Agency Start Date"])

# Below is to view the nba DataFrame and to get a summary of it
nyc_payroll
nyc_payroll.info()
nyc_payroll.describe( include="all" )

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Fiscal Year                 2000 non-null   int64         
 1   Payroll Number              2000 non-null   int64         
 2   Agency Name                 2000 non-null   object        
 3   Last Name                   1986 non-null   object        
 4   First Name                  1986 non-null   object        
 5   Mid Init                    1161 non-null   object        
 6   Agency Start Date           2000 non-null   datetime64[ns]
 7   Work Location Borough       2000 non-null   object        
 8   Title Description           2000 non-null   object        
 9   Leave Status as of June 30  2000 non-null   object        
 10  Base Salary                 2000 non-null   float64       
 11  Pay Basis                   2000 non-null   object      

<ipython-input-6-e81732bc0975>:15: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  nyc_payroll.describe( include="all" )


Fiscal Year  Payroll Number             Agency Name  Last Name  \
count   2000.000000     2000.000000                    2000       1986   
unique          NaN             NaN                      65       1620   
top             NaN             NaN  DEPT OF ED PEDAGOGICAL  RODRIGUEZ   
freq            NaN             NaN                     369         22   
first           NaN             NaN                     NaN        NaN   
last            NaN             NaN                     NaN        NaN   
mean    2022.500000      577.804000                     NaN        NaN   
std        0.500125      304.283239                     NaN        NaN   
min     2022.000000        2.000000                     NaN        NaN   
25%     2022.000000      300.000000                     NaN        NaN   
50%     2022.500000      742.000000                     NaN        NaN   
75%     2023.000000      747.000000                     NaN        NaN   
max     2023.000000      996.000000                     NaN        NaN   

       First Name Mid Init    Agency Start Date Work Location Borough  \
count        1986     1161                 2000                  2000   
unique       1162       25                 1170                     9   
top       MICHAEL        A  2020-01-01 00:00:00             MANHATTAN   
freq           23      165                   23                  1446   
first         NaN      NaN  1967-09-06 00:00:00                   NaN   
last          NaN      NaN  2023-06-12 00:00:00                   NaN   
mean          NaN      NaN                  NaN                   NaN   
std           NaN      NaN                  NaN                   NaN   
min           NaN      NaN                  NaN                   NaN   
25%           NaN      NaN                  NaN                   NaN   
50%           NaN      NaN                  NaN                   NaN   
75%           NaN      NaN                  NaN                   NaN   
max           NaN      NaN                  NaN                   NaN   

           Title Description Leave Status as of June 30    Base Salary  \
count                   2000                       2000    2000.000000   
unique                   272                          5            NaN   
top     TEACHER- PER SESSION                     ACTIVE            NaN   
freq                     339                       1613            NaN   
first                    NaN                        NaN            NaN   
last                     NaN                        NaN            NaN   
mean                     NaN                        NaN   49809.525245   
std                      NaN                        NaN   47915.863533   
min                      NaN                        NaN       1.000000   
25%                      NaN                        NaN      33.180000   
50%                      NaN                        NaN   46325.000000   
75%                      NaN                        NaN   92073.000000   
max                      NaN                        NaN  233094.000000   

        Pay Basis  Regular Hours  Regular Gross Paid     OT Hours  \
count        2000    2000.000000         2000.000000  2000.000000   
unique          4            NaN                 NaN          NaN   
top     per Annum            NaN                 NaN          NaN   
freq         1200            NaN                 NaN          NaN   
first         NaN            NaN                 NaN          NaN   
last          NaN            NaN                 NaN          NaN   
mean          NaN     616.458680        43918.969815    70.948525   
std           NaN     868.303506        43808.235364   191.727728   
min           NaN     -80.000000        -4174.920000   -12.250000   
25%           NaN       0.000000         3209.110000     0.000000   
50%           NaN       0.000000        31567.475000     0.000000   
75%           NaN    1820.000000        80958.612500     0.000000   
max           

In [ ]:
nyc_payroll

Fiscal Year  Payroll Number                     Agency Name Last Name  \
0            2022             740   DEPARTMENT OF EDUCATION ADMIN     DAVIS   
1            2022             746    DEPT OF ED PER DIEM TEACHERS      ZHOU   
2            2022             747  DEPT OF ED PER SESSION TEACHER     LOPEZ   
3            2022             742          DEPT OF ED PEDAGOGICAL   KENNEDY   
4            2022             742          DEPT OF ED PEDAGOGICAL   LICHVAR   
...           ...             ...                             ...       ...   
1995         2023             300  BOARD OF ELECTION POLL WORKERS    GARNER   
1996         2023             742          DEPT OF ED PEDAGOGICAL  FIGUEROA   
1997         2023             747  DEPT OF ED PER SESSION TEACHER   DIFORTE   
1998         2023             747  DEPT OF ED PER SESSION TEACHER    BIONDO   
1999         2023             827        DEPARTMENT OF SANITATION     ARIAS   

     First Name Mid Init Agency Start Date Work Location Borough  \
0         OSCAR      NaN        2021-07-06                 BRONX   
1           JUN      NaN        2020-07-01             MANHATTAN   
2       YERIBEL        E        2001-12-11             MANHATTAN   
3          VERA        R        2016-04-08             MANHATTAN   
4        AMANDA        C        2007-02-01             MANHATTAN   
...         ...      ...               ...                   ...   
1995   GEOFFREY        O        2020-01-01             MANHATTAN   
1996    JESSICA        A        2006-02-08             MANHATTAN   
1997   DANIELLE        N        2017-09-05             MANHATTAN   
1998   STEFANIE      NaN        2011-10-03             MANHATTAN   
1999    ELPIDIO      NaN        2022-07-11             MANHATTAN   

                                      Title Description  \
0     SUBSTANCE ABUSE PREVENTION & INTERVENTION SPEC...   
1                                    TEACHER-GENERAL ED   
2                                  TEACHER- PER SESSION   
3                                               TEACHER   
4                             TEACHER SPECIAL EDUCATION   
...                                                 ...   
1995                                    ELECTION WORKER   
1996                                 TEACHER ASSIGNED A   
1997                               TEACHER- PER SESSION   
1998                               TEACHER- PER SESSION   
1999                                  SANITATION WORKER   

     Leave Status as of June 30  Base Salary  Pay Basis  Regular Hours  \
0                        CEASED     53820.00  per Annum          805.0   
1                        ACTIVE       110.29    per Day            0.0   
2                        ACTIVE        33.18    per Day            0.0   
3                        ACTIVE     79687.00  per Annum            0.0   
4                        ACTIVE    109852.00  per Annum            0.0   
...                         ...          ...        ...            ...   
1995                     ACTIVE         1.00   per Hour            0.0   
1996                     ACTIVE    109852.00  per Annum            0.0   
1997                     ACTIVE        33.18    per Day            0.0   
1998                     ACTIVE        33.18    per Day            0.0   
1999                     ACTIVE     44064.00  per Annum         2000.0   

      Regular Gross Paid  OT Hours  Total OT Paid  Total Other Pay  
0               24469.41       0.0           0.00             0.00  
1                3387.59       0.0           0.00             0.00  
2               20242.24       0.0           0.00           474.32  
3               74904.27       0.0           0.00         -4474.44  
4              109852.08       0.0           0.00          5380.44  
...                  ...       ...            ...              ...  
1995              928.00       0.0           0.00             0.00  
1996           109852.08       0.0           0.00             0.00  
1997             1824

In [ ]:
# HW3 Q1 - Select “First Name”, “Last Name”, “Base Salary”, and “Total OT Paid”, then sort these selected variables by “Base Salary” in descending order and display the top 10 entries.

import pandas as pd
from google.colab import data_table
data_table.enable_dataframe_formatter()
nyc_payroll = pd.read_csv('https://bcdanl.github.io/data/nyc_payroll_2022_2023.csv',
                          parse_dates = ["Agency Start Date"])
nyc_payroll

Fiscal Year  Payroll Number                     Agency Name Last Name  \
0            2022             740   DEPARTMENT OF EDUCATION ADMIN     DAVIS   
1            2022             746    DEPT OF ED PER DIEM TEACHERS      ZHOU   
2            2022             747  DEPT OF ED PER SESSION TEACHER     LOPEZ   
3            2022             742          DEPT OF ED PEDAGOGICAL   KENNEDY   
4            2022             742          DEPT OF ED PEDAGOGICAL   LICHVAR   
...           ...             ...                             ...       ...   
1995         2023             300  BOARD OF ELECTION POLL WORKERS    GARNER   
1996         2023             742          DEPT OF ED PEDAGOGICAL  FIGUEROA   
1997         2023             747  DEPT OF ED PER SESSION TEACHER   DIFORTE   
1998         2023             747  DEPT OF ED PER SESSION TEACHER    BIONDO   
1999         2023             827        DEPARTMENT OF SANITATION     ARIAS   

     First Name Mid Init Agency Start Date Work Location Borough  \
0         OSCAR      NaN        2021-07-06                 BRONX   
1           JUN      NaN        2020-07-01             MANHATTAN   
2       YERIBEL        E        2001-12-11             MANHATTAN   
3          VERA        R        2016-04-08             MANHATTAN   
4        AMANDA        C        2007-02-01             MANHATTAN   
...         ...      ...               ...                   ...   
1995   GEOFFREY        O        2020-01-01             MANHATTAN   
1996    JESSICA        A        2006-02-08             MANHATTAN   
1997   DANIELLE        N        2017-09-05             MANHATTAN   
1998   STEFANIE      NaN        2011-10-03             MANHATTAN   
1999    ELPIDIO      NaN        2022-07-11             MANHATTAN   

                                      Title Description  \
0     SUBSTANCE ABUSE PREVENTION & INTERVENTION SPEC...   
1                                    TEACHER-GENERAL ED   
2                                  TEACHER- PER SESSION   
3                                               TEACHER   
4                             TEACHER SPECIAL EDUCATION   
...                                                 ...   
1995                                    ELECTION WORKER   
1996                                 TEACHER ASSIGNED A   
1997                               TEACHER- PER SESSION   
1998                               TEACHER- PER SESSION   
1999                                  SANITATION WORKER   

     Leave Status as of June 30  Base Salary  Pay Basis  Regular Hours  \
0                        CEASED     53820.00  per Annum          805.0   
1                        ACTIVE       110.29    per Day            0.0   
2                        ACTIVE        33.18    per Day            0.0   
3                        ACTIVE     79687.00  per Annum            0.0   
4                        ACTIVE    109852.00  per Annum            0.0   
...                         ...          ...        ...            ...   
1995                     ACTIVE         1.00   per Hour            0.0   
1996                     ACTIVE    109852.00  per Annum            0.0   
1997                     ACTIVE        33.18    per Day            0.0   
1998                     ACTIVE        33.18    per Day            0.0   
1999                     ACTIVE     44064.00  per Annum         2000.0   

      Regular Gross Paid  OT Hours  Total OT Paid  Total Other Pay  
0               24469.41       0.0           0.00             0.00  
1                3387.59       0.0           0.00             0.00  
2               20242.24       0.0           0.00           474.32  
3               74904.27       0.0           0.00         -4474.44  
4              109852.08       0.0           0.00          5380.44  
...                  ...       ...            ...              ...  
1995              928.00       0.0           0.00             0.00  
1996           109852.08       0.0           0.00             0.00  
1997             1824

In [ ]:
nyc_payroll2.nlargest(10, 'Base Salary', keep = 'all') # This line will show the top 10 entries in terms of Base Salary in the updated Data Frame nyc_Payroll2

First Name  Last Name  Base Salary  Total OT Paid
433     MICHAEL    BIMONTE     233094.0            0.0
61       BEDROS  BOODANIAN     215861.0            0.0
1378     ARVIND     SOHONI     215000.0            0.0
1469     ANDREW     KAPLAN     195000.0            0.0
73     VICTORIA       RICE     187530.0            0.0
858      JOSEPH   PILEWSKI     180450.0            0.0
1379     SAMUEL      DAVID     180000.0            0.0
425      KARINA    JIMENEZ     178000.0            0.0
74     JOYCELYN       NEDD     177113.0            0.0
352    JEANETTE     VARGAS     176947.0            0.0

I updated the nyc_payroll DataFrame by creating a new DataFrame nyc_payroll2 that shows only the columns First Name, Last Name, Base Salary, and Total OT Paid. From there, I used the nlargest function to show the top 10 entries in terms of salary.

In [ ]:
# HW3 Q2 - Create a new DataFrame that includes employees from the “DEPARTMENT OF EDUCATION ADMIN” agency where the variables are “First Name”, “Last Name”, “Title Description”, “Base Salary”, and “Total OT Paid”. Additionally, include a new variable “Total Compensation” which is the sum of “Base Salary” and “Total OT Paid”.

nyc_payroll3 = nyc_payroll[['First Name', 'Last Name', "Title Description", 'Base Salary', 'Total OT Paid']] # This line will List the selected columns
nyc_payroll3["Total Compensation"] = nyc_payroll3["Base Salary"] + nyc_payroll3["Total OT Paid"] # This line is to create a new "Total Compensation Column" that adds Base Salary with Total OT Paid
nyc_payroll3.loc[ "DEPARTMENT OF EDUCATION ADMIN" ] # This line will locate Department of Education Admin employees only in the index



<ipython-input-47-d51f359cd59c>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nyc_payroll3["Total Compensation"] = nyc_payroll3["Base Salary"] + nyc_payroll3["Total OT Paid"]


First Name     Last Name  \
Agency Name                                              
DEPARTMENT OF EDUCATION ADMIN      OSCAR         DAVIS   
DEPARTMENT OF EDUCATION ADMIN      JENNA    VARNAUSKAS   
DEPARTMENT OF EDUCATION ADMIN  CHRISTINA     GRANQUIST   
DEPARTMENT OF EDUCATION ADMIN     SANDRA         QUIAH   
DEPARTMENT OF EDUCATION ADMIN    SHATARA          EARL   
...                                  ...           ...   
DEPARTMENT OF EDUCATION ADMIN   DEISMOND          ROSA   
DEPARTMENT OF EDUCATION ADMIN     SERIDE          READ   
DEPARTMENT OF EDUCATION ADMIN      ELISA  CORADO DERAS   
DEPARTMENT OF EDUCATION ADMIN     YAZMYN        YEARDE   
DEPARTMENT OF EDUCATION ADMIN   FOLASADE     AKINTUNDE   

                                                               Title Description  \
Agency Name                                                                        
DEPARTMENT OF EDUCATION ADMIN  SUBSTANCE ABUSE PREVENTION & INTERVENTION SPEC...   
DEPARTMENT OF EDUCATION ADMIN                             OCCUPATIONAL THERAPIST   
DEPARTMENT OF EDUCATION ADMIN                                 PHYSICAL THERAPIST   
DEPARTMENT OF EDUCATION ADMIN                                COMMUNITY ASSOCIATE   
DEPARTMENT OF EDUCATION ADMIN                                PROCUREMENT ANALYST   
...                                                                          ...   
DEPARTMENT OF EDUCATION ADMIN                              COMMUNITY COORDINATOR   
DEPARTMENT OF EDUCATION ADMIN                              COMMUNITY COORDINATOR   
DEPARTMENT OF EDUCATION ADMIN                                COMMUNITY ASSISTANT   
DEPARTMENT OF EDUCATION ADMIN  SUBSTANCE ABUSE PREVENTION & INTERVENTION SPEC...   
DEPARTMENT OF EDUCATION ADMIN                                 PHYSICAL THERAPIST   

                               Base Salary  Total OT Paid  Total Compensation  
Agency Name                                                                    
DEPARTMENT OF EDUCATION ADMIN      53820.0           0.00            53820.00  
DEPARTMENT OF EDUCATION ADMIN      75381.0           0.00            75381.00  
DEPARTMENT OF EDUCATION ADMIN      80270.0           0.00            80270.00  
DEPARTMENT OF EDUCATION ADMIN      55323.0           0.00            55323.00  
DEPARTMENT OF EDUCATION ADMIN      62055.0           0.00            62055.00  
...                                    ...            ...                 ...  
DEPARTMENT OF EDUCATION ADMIN      82170.0        1602.93            83772.93  
DEPARTMENT OF EDUCATION ADMIN      67834.0         639.27            68473.27  
DEPARTMENT OF EDUCATION ADMIN      40765.0           0.00            40765.00  
DEPARTMENT OF EDUCATION ADMIN      58811.0           0.00            58811.00  
DEPARTMENT OF EDUCATION ADMIN      80730.0           0.00            80730.00  

[73 rows x 6 columns]

I first used the .set_index function to set the index to Agency Name. I then created a new DataFrame nyc_payroll3 with selected columns. I then used vectorized operations to add together the base salary and total OT paid columns to a new collumn called total compensation. Finally I located just Department of Education Admin employees using the .loc function on the index column.  

In [ ]:
import pandas as pd
from google.colab import data_table
data_table.enable_dataframe_formatter()
nyc_payroll = pd.read_csv('https://bcdanl.github.io/data/nyc_payroll_2022_2023.csv',
                          parse_dates = ["Agency Start Date"])

nyc_payroll

# HW3 Q3 - Using set_index(), change the DataFrame’s index to “Last Name”, then locate the data for a specific last name, say “Doe”, and display their “Agency Name”, “Base Salary”, and “Total OT Paid”.

nyc_payroll4 = nyc_payroll[['Last Name', 'Agency Name', 'Base Salary', 'Total OT Paid']] # Create a new DF with selected rows
nyc_payroll4.set_index("Last Name", inplace = True) # Set new DF index to Last Name
nyc_payroll4.loc[ "DAVIS" ] # Locate last name "Davis" in the index

Agency Name  Base Salary  Total OT Paid
Last Name                                                            
DAVIS       DEPARTMENT OF EDUCATION ADMIN     53820.00           0.00
DAVIS               NYC HOUSING AUTHORITY     32950.00        2332.60
DAVIS         HRA/DEPT OF SOCIAL SERVICES     81452.00        6253.86
DAVIS             BRONX DISTRICT ATTORNEY    160000.00           0.00
DAVIS      DEPT OF ED PER SESSION TEACHER        33.18           0.00
DAVIS         HRA/DEPT OF SOCIAL SERVICES     61178.00           0.00
DAVIS      DEPT OF ENVIRONMENT PROTECTION     38823.00         225.99

As I was trying different things, I realized the main nyc_payroll DF had weird column names. I reset it by importing the information again and setting the original DF.

In [ ]:
# HW3 Q4 - Find the 5 employees with the highest “Regular Gross Paid” and calculate their average “OT Hours”. Also, reset the index if you have changed it previously.

nyc_payroll.reset_index() # Reset index
nyc_payroll5 = nyc_payroll[['Last Name', 'First Name', 'Regular Gross Paid', 'OT Hours']] # Create new DF with selected rows
nyc_payroll5.nlargest(5, 'Regular Gross Paid') # Find 5 highest regular gross paid employees


Last Name First Name  Regular Gross Paid  OT Hours
61    BOODANIAN     BEDROS           215269.60       0.0
1378     SOHONI     ARVIND           193479.52       0.0
73         RICE   VICTORIA           186164.52       0.0
1469     KAPLAN     ANDREW           178356.14       0.0
1438      NOSSA     ROBERT           176719.36      32.5

In [ ]:
nyc_payroll5.mean() # Calculate the average OT Hours for all employees

<ipython-input-106-be7c015c8e34>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  nyc_payroll5.mean() # Calculate the average OT Hours


Regular Gross Paid    43918.969815
OT Hours                 70.948525
dtype: float64

In [ ]:
nyc_payroll5.nlargest(5, 'Regular Gross Paid').mean() # Calculate the average OT Hours for the top 5 highest regular gross paid employees

<ipython-input-109-31deb29f44a3>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  nyc_payroll5.nlargest(5, 'Regular Gross Paid').mean()


Regular Gross Paid    189997.828
OT Hours                   6.500
dtype: float64

I was unsure if you wanted us to average just those 5 employees or the entire list of employees OT hours. I wasn't sure how to do only the 5 employees listed. I did both just in case.

In [64]:
import pandas as pd
from google.colab import data_table
data_table.enable_dataframe_formatter()
nyc_payroll = pd.read_csv('https://bcdanl.github.io/data/nyc_payroll_2022_2023.csv',
                          parse_dates = ["Agency Start Date"])

nyc_payroll

# HW3 Q5 - Sort the DataFrame by “Fiscal Year” and “Total Other Pay” in descending order, then set “First Name” as the index and use the loc accessor to retrieve the “Total Other Pay” for a specific first name, say “John”.

nyc_payroll8 = nyc_payroll[['First Name', 'Last Name', 'Fiscal Year', 'Total Other Pay']]
nyc_payroll9 = nyc_payroll8.sort_values(by=["Fiscal Year", "Total Other Pay"],
                        ascending = False)
nyc_payroll9.set_index('First Name', inplace=True)
nyc_payroll9.loc[
    "ANDREW",
    ["Fiscal Year", "Total Other Pay"]
]




Fiscal Year  Total Other Pay
First Name                              
ANDREW             2023          3579.58
ANDREW             2023          1744.11
ANDREW             2023             0.00
ANDREW             2023         -4242.82
ANDREW             2022         13650.29
ANDREW             2022          5774.37
ANDREW             2022          5655.70
ANDREW             2022          2235.49
ANDREW             2022             0.00
ANDREW             2022             0.00

I tried everything to get them to get "Total Other Pay" to sort in descending order but was unable to figure it out.

UPDATE: Brian and I worked this out and it is working now.

In [65]:
import pandas as pd
from google.colab import data_table
data_table.enable_dataframe_formatter()
nyc_payroll = pd.read_csv('https://bcdanl.github.io/data/nyc_payroll_2022_2023.csv',
                          parse_dates = ["Agency Start Date"])

nyc_payroll

# HW3 Q6 - Select employees who have "OT Hours" greater than 0, calculate their "OT Rate", and then find the employee with the highest "OT Rate". Display their full name and OT Rate

# Boolean Series of whether or not 'OT Hours' is greater than 0.
pos_OT_hours = nyc_payroll['OT Hours'] > 0

# Keeping only the observations with pos_OT_hours == True
with_ot = nyc_payroll[ pos_OT_hours ]

# Calculating OT Rate
with_ot["OT Rate"] = with_ot["Total OT Paid"] / with_ot["OT Hours"]

#Find Highest OT Rate
with_ot.sort_values(["OT Rate"],
                    ascending = False)

# Display their full name and OT Rate
nyc_payroll7 = with_ot[['First Name','Last Name' , 'OT Rate']]

# Find highest OT Rate
nyc_payroll7.nlargest(1, 'OT Rate')



<ipython-input-65-66d0b540018d>:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  with_ot["OT Rate"] = with_ot["Total OT Paid"] / with_ot["OT Hours"]


First Name Last Name     OT Rate
800     DANIEL   DEBONIS  174.115769

The running code is now displaying the employee with the highest OT Rate by full name. The individuals name is Daniel Debonis and his OT Rate is 174.116.

In [ ]:
# HW3 Q7 - Sort the DataFrame first by “Work Location Borough” alphabetically, and then by “Total Compensation” (sum of “Base Salary” and “Total OT Paid”) in descending order within each borough.

In [73]:
import pandas as pd
from google.colab import data_table
data_table.enable_dataframe_formatter()
nyc_payroll = pd.read_csv('https://bcdanl.github.io/data/nyc_payroll_2022_2023.csv',
                          parse_dates = ["Agency Start Date"])

nyc_payroll

# Create "Total Compensation" column
nyc_payroll['Total Compensation'] = nyc_payroll['Base Salary'] + nyc_payroll['Total OT Paid']

# Sort DF by "Work Location Borough" alphabetically
nyc_payroll10 = nyc_payroll[['First Name', 'Last Name', 'Base Salary', 'Total OT Paid', 'Work Location Borough', 'Total Compensation' ]]
nyc_payroll11 = nyc_payroll10.sort_values(["Work Location Borough", "Total Compensation"],
                    ascending = [True, False])
nyc_payroll11

First Name Last Name  Base Salary  Total OT Paid Work Location Borough  \
1172    DOMINIC  ROBINSON    118056.00       68725.63                 BRONX   
253      DONALD  HANRATTY    170000.00           0.00                 BRONX   
1182    JEFFREY     DAVIS    160000.00           0.00                 BRONX   
602     ANTHONY     RUMPH    105062.00       49706.74                 BRONX   
307        RYAN     WEBER     92073.00       61087.56                 BRONX   
...         ...       ...          ...            ...                   ...   
688       KELLY    HAMLIN     65640.00           0.00              SULLIVAN   
1137      KAREN   COLLINS     68962.00           1.10                ULSTER   
123      THOMAS  FOULKROD     67758.00           0.00                ULSTER   
1797     LEONID  TSVITMAN     57593.00        3643.47           WESTCHESTER   
1260     ROBERT  HOULIHAN       440.16        8630.28           WESTCHESTER   

      Total Compensation  
1172           186781.63  
253            170000.00  
1182           160000.00  
602            154768.74  
307            153160.56  
...                  ...  
688             65640.00  
1137            68963.10  
123             67758.00  
1797            61236.47  
1260             9070.44  

[2000 rows x 6 columns]

The code above now shows each employee, sorted by Work Location Borough alphbetically and Total Campensation from largest to smallest within each borough.